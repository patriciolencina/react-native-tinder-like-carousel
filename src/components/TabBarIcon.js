import React, {useRef} from 'react';
import {TouchableWithoutFeedback, View, Animated, Easing} from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';
import styles from '../screens/TabNavigator/TabBar/styles';

const TabBarIcon = ({focused, onPress, index, routeName}) => {
  const animatedValue = useRef(new Animated.Value(0)).current;

  if (focused) {
    Animated.timing(animatedValue, {
      toValue: 1,
      duration: 200,
      easing: Easing.ease,
    }).start();
  } else {
    Animated.timing(animatedValue, {
      toValue: 0,
      duration: 50,
      easing: Easing.ease,
    }).start();
  }

  let icon = '';

  switch (index) {
    case 0:
      icon = 'ios-home';
      break;

    case 1:
      icon = 'ios-heart';
      break;
  }
  const translateXObject = {
    translateX: animatedValue.interpolate({
      inputRange: [0, 1],
      outputRange: [0, index === 0 ? 2 : -2],
    }),
  };
  return (
    <TouchableWithoutFeedback onPress={() => onPress(routeName)}>
      <View style={[styles.wrapIcon]}>
        <Animated.View
          style={[
            styles.icon,
            {
              transform: [
                {
                  translateY: animatedValue.interpolate({
                    inputRange: [0, 1],
                    outputRange: [0, -15],
                  }),
                },
                translateXObject,
              ],
            },
          ]}>
          <Icon name={icon} color="black" size={34} />
        </Animated.View>
        <Animated.Text
          style={[
            styles.textStyle,
            {
              transform: [
                {
                  translateY: animatedValue.interpolate({
                    inputRange: [0, 1],
                    outputRange: [0, -40],
                  }),
                },
                translateXObject,
              ],
            },
          ]}>
          {routeName}
        </Animated.Text>
      </View>
    </TouchableWithoutFeedback>
  );
};

export default TabBarIcon;
