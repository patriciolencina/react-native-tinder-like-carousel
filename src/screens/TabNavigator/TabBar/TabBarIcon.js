import * as React from 'react';
import {TouchableWithoutFeedback, View, Animated, Easing} from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';
import styles from './styles';
export default class TabBarIcon extends React.PureComponent {
  constructor(props) {
    super(props);
    this.animatedValue = new Animated.Value(0);
  }
  handleAnimation = () => {
    if (this.props.focused) {
      Animated.timing(this.animatedValue, {
        toValue: 1,
        duration: 200,
        easing: Easing.ease,
      }).start();
    } else {
      Animated.timing(this.animatedValue, {
        toValue: 0,
        duration: 50,
        easing: Easing.ease,
      }).start();
    }
  };
  onSelect = routeName => this.props.onPress(routeName);

  render() {
    this.handleAnimation();

    const {index, routeName} = this.props;
    let icon = '';

    switch (index) {
      case 0:
        icon = 'ios-home';
        break;

      case 1:
        icon = 'ios-heart';
        break;
    }
    const translateXObject = {
      translateX: this.animatedValue.interpolate({
        inputRange: [0, 1],
        outputRange: [0, index < 2 ? 2 : index === 2 ? 0 : -2],
      }),
    };
    return (
      <TouchableWithoutFeedback onPress={() => this.onSelect(routeName)}>
        <View style={[styles.wrapIcon]}>
          <Animated.View
            style={[
              styles.icon,
              {
                transform: [
                  {
                    translateY: this.animatedValue.interpolate({
                      inputRange: [0, 1],
                      outputRange: [0, -15],
                    }),
                  },
                  translateXObject,
                ],
              },
            ]}>
            <Icon name={icon} color="black" size={34} />
          </Animated.View>
          <Animated.Text
            style={[
              styles.textStyle,
              {
                transform: [
                  {
                    translateY: this.animatedValue.interpolate({
                      inputRange: [0, 1],
                      outputRange: [0, -40],
                    }),
                  },
                  translateXObject,
                ],
              },
            ]}>
            {routeName}
          </Animated.Text>
        </View>
      </TouchableWithoutFeedback>
    );
  }
}
