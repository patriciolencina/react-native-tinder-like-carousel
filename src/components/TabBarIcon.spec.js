import React from 'react';

import {shallow, ShallowWrapper} from 'enzyme';
import TabBarIcon from './TabBarIcon';

const createTestProps = (props: Object) => ({
  index: 0,
  ...props,
});

describe('TabBarIcon', () => {
  describe('rendering', () => {
    let wrapper: ShallowWrapper;
    let props: any; // use type "any" to opt-out of type-checking
    beforeEach(() => {
      props = createTestProps({});
      wrapper = shallow(<TabBarIcon {...props} />); // no compile-time error
    });

    it('should render', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
