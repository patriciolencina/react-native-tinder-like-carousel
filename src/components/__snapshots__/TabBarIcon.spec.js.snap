// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBarIcon rendering should render 1`] = `
<TouchableWithoutFeedback
  onPress={[Function]}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
        },
      ]
    }
  >
    <AnimatedComponent
      style={
        Array [
          undefined,
          Object {
            "transform": Array [
              Object {
                "translateY": 0,
              },
              Object {
                "translateX": 0,
              },
            ],
          },
        ]
      }
    >
      <Icon
        allowFontScaling={false}
        color="black"
        name="ios-home"
        size={34}
      />
    </AnimatedComponent>
    <AnimatedComponent
      style={
        Array [
          Object {
            "color": "black",
            "fontSize": 9,
            "marginTop": 20,
            "textAlign": "center",
          },
          Object {
            "transform": Array [
              Object {
                "translateY": 0,
              },
              Object {
                "translateX": 0,
              },
            ],
          },
        ]
      }
    />
  </View>
</TouchableWithoutFeedback>
`;
